### PR TITLE
Fixed https://github.com/couchbase/couchbase-lite-android/issues/898

### PR DESCRIPTION
**Pull request by hideki**
Push replication failing with large data sets

- Solution: Implement pause logic into PusherInternal. In case total count of Batcher and pendingSequences sizes are more than 200, push replicator pauses to add changes into Batcher. This is same approach with PullerInternal pause logic.

### CODE REVIEW

File: src/main/java/com/couchbase/lite/replicator/PusherInternal.java [hunk 1]
```diff
@@ -209,23 +271,25 @@ public void beginReplicating() {
...
             //       submit changes  into inbox, there are chance both inbox and pendingFutures are
             //       empty.
             submitRevisions(changes);
         } else {
-            Log.d(Log.TAG_SYNC, "%s: No changes since %d", this, lastSequenceLong);
+            Log.d(TAG, "%s: No changes since %d", this, lastSequenceLong);
         }
 
         // process queued changes by `changed()` callback
         synchronized (changesLock) {
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PusherInternal.java]: I think synchronize here doesn't need anymore as the code will be run in the support work executor.

File: src/main/java/com/couchbase/lite/replicator/PusherInternal.java [hunk 2]
```diff
@@ -209,23 +271,25 @@ public void beginReplicating() {
...
+        Log.d(TAG, "%s: Getting changes since %d", this, lastSequenceLong);
         final RevisionList changes = db.changesSince(lastSequenceLong, options, filter, filterParams);
         if (changes.size() > 0) {
-            Log.d(Log.TAG_SYNC, "%s: Queuing %d changes since %d", this, changes.size(), lastSequenceLong);
+            Log.d(TAG, "%s: Queuing %d changes since %d", this, changes.size(), lastSequenceLong);
             // NOTE: Needs to submit changes into inbox from RemoteRequest thread for beginReplication.
             //       RemoteRequest thread is observed by pendingFuture, if using other thread to
             //       submit changes  into inbox, there are chance both inbox and pendingFutures are
             //       empty.
             submitRevisions(changes);
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PusherInternal.java]: Does the NOTE above this line is still applicable?

File: src/main/java/com/couchbase/lite/replicator/PusherInternal.java [hunk 3]
```diff
@@ -98,14 +115,48 @@ public void setCreateTarget(boolean createTarget) {
...
-        Log.d(Log.TAG_SYNC, "%s STOPPING...", toString());
+        Log.d(TAG, "%s STOPPING...", toString());
 
         stopObserving();
 
+        // Awake thread if it is wait for pause
+        setPaused(false);
+
+        // shutdown supportExecutor immediately, does not add any more tasks.
+        terminateSupportExecutor();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PusherInternal.java]: Could this cause race condition between terminateSupportExecutor() and adding tasks to the executor? The race condition might be ok if adding the tasks to the work executor after being terminated has no side effect.

Comment (hideki) [inline src/main/java/com/couchbase/lite/replicator/PusherInternal.java]: As you wrote, I believe race-condition does not cause the problem. Once executor is shutdown, any new tasks are not processed.

File: src/main/java/com/couchbase/lite/replicator/PusherInternal.java (hunk 2 above)
Comment (hideki) [inline src/main/java/com/couchbase/lite/replicator/PusherInternal.java]: Thank you for catching this. This is still applicable.

**Comment by hideki** [comment]
@pasin, Could you please review again?

Changes
- make `beginReplicating()` not call from `ExecutorService` to avoid potential problem
- Instead, make minimum thread-pool size of remote request executor to 2

**Comment by pasin** [comment]
We should create a ticket to rethink about the replicator work executor and the way that waits for pending future and etc.

**Pull request merged**